File: week12/problem-set/prob1.py
"""Prob1 example.

This file belongs to the week 12 geometry and compression examples. It uses euclidean_distance, merge_sort, merge, closest_pair, and related helpers to illustrate the main idea with a small runnable example.
"""

import math
import matplotlib.pyplot as plt

def euclidean_distance(p1, p2):
    """Return the straight-line distance between two points.

    Args:
        p1 (tuple): The first point that defines the reference line or distance check.
        p2 (tuple): The second point that defines the reference line or distance check.

    Returns:
        float: The numeric result produced by the algorithm.
    """
    return math.sqrt((p1[0] - p2[0]) ** 2 + (p1[1] - p2[1]) ** 2)


def merge_sort(points, key):
    """Sort the list using merge sort.

    Args:
        points (list): The points used by the geometry algorithm.
        key (int): The key input used by this function.

    Returns:
        list: The list produced by the algorithm.
    """
    if len(points) <= 1:
        return points
    middle = len(points) // 2
    left = merge_sort(points[:middle], key)
    right = merge_sort(points[middle:], key)

    return merge(left, right, key)


def merge(left, right, key):
    """Merge two sorted halves into a single sorted result.

    Args:
        left (object): The left boundary or left-side data used by the algorithm.
        right (object): The right boundary or right-side data used by the algorithm.
        key (int): The key input used by this function.

    Returns:
        list: The list produced by the algorithm.
    """
    merged = []
    i, j = 0, 0
    while i < len(left) and j < len(right):
        if left[i][key] < right[j][key]:
            merged.append(left[i])
            i += 1
        else:
            merged.append(right[j])
            j += 1

    while i < len(left):
        merged.append(left[i])
        i += 1

    while j < len(right):
        merged.append(right[j])
        j += 1

    return merged


def closest_pair(points, sorted_y):
    """Return the closest pair of points using divide and conquer.

    Args:
        points (list): The points used by the geometry algorithm.
        sorted_y (list): The same points sorted by their y-coordinate.

    Returns:
        tuple: The tuple of values produced by the algorithm.
    """
    if len(points) <= 3:
        min_distance = float('inf')
        min_pair = None
        for i in range(len(points)):
            for j in range(i + 1, len(points)):
                distance = euclidean_distance(points[i], points[j])
                if distance < min_distance:
                    min_distance = distance
                    min_pair = (points[i], points[j])
        return min_pair, min_distance

    mid = len(points) // 2
    left_points = points[:mid]
    right_points = points[mid:]

    left_sorted_y = [p for p in sorted_y if p in left_points]
    right_sorted_y = [p for p in sorted_y if p in right_points]

    left_pair, left_distance = closest_pair(left_points, left_sorted_y)
    right_pair, right_distance = closest_pair(right_points, right_sorted_y)
    min_pair = left_pair if left_distance < right_distance else right_pair
    min_distance = min(left_distance, right_distance)

    strip_points = [p for p in sorted_y if abs(p[0] - points[mid][0]) < min_distance]

    for i in range(len(strip_points)):
        for j in range(i + 1, len(strip_points)):
            if strip_points[j][1] - strip_points[i][1] >= min_distance:
                break
            distance = euclidean_distance(strip_points[i], strip_points[j])
            if distance < min_distance:
                min_distance = distance
                min_pair = (strip_points[i], strip_points[j])

    return min_pair, min_distance


def main():
    """Run the small demonstration for this teaching example.

    Returns:
        None: This function updates data or prints results and does not return a value.
    """
    points = [(2, 3), (12, 30), (40, 50), (5, 1), (12, 10), (3, 4), (30, 25)]
    sorted_points = merge_sort(points, 0)
    sorted_y = merge_sort(points, 1)
    min_pair, min_distance = closest_pair(sorted_points, sorted_y)
    print(f"Closest pair: {min_pair}")
    print(f"Minimum distance: {min_distance}")

    x = [point[0] for point in points]
    y = [point[1] for point in points]

    x_res = [point[0] for point in min_pair]
    y_res = [point[1] for point in min_pair]
    col = ['red' if point in min_pair else 'blue' for point in points ]
    plt.plot(x_res,y_res)
    plt.scatter(x,y,c = col)
    plt.show()


if __name__ == "__main__":
    main()

File: week12/problem-set/prob1.cpp
/**
 * @file prob1.cpp
 * @brief Console-based C++ version of the closest-pair-of-points example.
 */

#include <cmath>
#include <iostream>
#include <utility>
#include <vector>

using namespace std;

using Point = pair<double, double>;

/**
 * @brief Return the straight-line distance between two points.
 * @param p1 const Point& First point as (x, y).
 * @param p2 const Point& Second point as (x, y).
 * @return double Euclidean distance between p1 and p2.
 */
double euclidean_distance(const Point& p1, const Point& p2) {
    double dx = p1.first - p2.first;
    double dy = p1.second - p2.second;
    return sqrt(dx * dx + dy * dy);
}

/**
 * @brief Merge two sorted point lists using either x or y as the key.
 * @param left const vector<Point>& First sorted half.
 * @param right const vector<Point>& Second sorted half.
 * @param key int Coordinate index: 0 means x, 1 means y.
 * @return vector<Point> Merged sorted list.
 */
vector<Point> merge(const vector<Point>& left, const vector<Point>& right, int key) {
    vector<Point> merged;
    size_t i = 0;
    size_t j = 0;

    while (i < left.size() && j < right.size()) {
        double left_key = key == 0 ? left[i].first : left[i].second;
        double right_key = key == 0 ? right[j].first : right[j].second;
        if (left_key < right_key) {
            merged.push_back(left[i++]);
        } else {
            merged.push_back(right[j++]);
        }
    }
    merged.insert(merged.end(), left.begin() + static_cast<long>(i), left.end());
    merged.insert(merged.end(), right.begin() + static_cast<long>(j), right.end());
    return merged;
}

/**
 * @brief Sort points using merge sort and the selected coordinate as the key.
 * @param points const vector<Point>& Input point list.
 * @param key int Coordinate index: 0 means x, 1 means y.
 * @return vector<Point> Sorted copy of points.
 */
vector<Point> merge_sort(const vector<Point>& points, int key) {
    if (points.size() <= 1) {
        return points;
    }
    size_t middle = points.size() / 2;
    vector<Point> left(points.begin(), points.begin() + static_cast<long>(middle));
    vector<Point> right(points.begin() + static_cast<long>(middle), points.end());
    return merge(merge_sort(left, key), merge_sort(right, key), key);
}

/**
 * @brief Return the closest pair of points using divide and conquer.
 * @param points const vector<Point>& Points sorted by x-coordinate.
 * @param sorted_y const vector<Point>& Same points sorted by y-coordinate.
 * @return pair<pair<Point, Point>, double> Closest pair and the distance between them.
 */
pair<pair<Point, Point>, double> closest_pair(const vector<Point>& points, const vector<Point>& sorted_y) {
    if (points.size() <= 3) {
        double min_distance = 1e18;
        pair<Point, Point> min_pair;
        for (size_t i = 0; i < points.size(); ++i) {
            for (size_t j = i + 1; j < points.size(); ++j) {
                double distance = euclidean_distance(points[i], points[j]);
                if (distance < min_distance) {
                    min_distance = distance;
                    min_pair = {points[i], points[j]};
                }
            }
        }
        return {min_pair, min_distance};
    }

    size_t mid = points.size() / 2;
    vector<Point> left_points(points.begin(), points.begin() + static_cast<long>(mid));
    vector<Point> right_points(points.begin() + static_cast<long>(mid), points.end());

    vector<Point> left_sorted_y;
    vector<Point> right_sorted_y;
    for (const Point& point : sorted_y) {
        bool belongs_to_left = false;
        for (const Point& left_point : left_points) {
            if (left_point == point) {
                belongs_to_left = true;
                break;
            }
        }
        if (belongs_to_left) {
            left_sorted_y.push_back(point);
        } else {
            right_sorted_y.push_back(point);
        }
    }

    auto left_result = closest_pair(left_points, left_sorted_y);
    auto right_result = closest_pair(right_points, right_sorted_y);
    auto best_result = left_result.second < right_result.second ? left_result : right_result;

    vector<Point> strip_points;
    double middle_x = points[mid].first;
    for (const Point& point : sorted_y) {
        if (fabs(point.first - middle_x) < best_result.second) {
            strip_points.push_back(point);
        }
    }

    for (size_t i = 0; i < strip_points.size(); ++i) {
        for (size_t j = i + 1; j < strip_points.size(); ++j) {
            if (strip_points[j].second - strip_points[i].second >= best_result.second) {
                break;
            }
            double distance = euclidean_distance(strip_points[i], strip_points[j]);
            if (distance < best_result.second) {
                best_result = {{strip_points[i], strip_points[j]}, distance};
            }
        }
    }
    return best_result;
}

/**
 * @brief Run the closest-pair demonstration.
 * @return int Exit status code returned to the operating system.
 */
int main() {
    vector<Point> points = {{2, 3}, {12, 30}, {40, 50}, {5, 1}, {12, 10}, {3, 4}, {30, 25}};
    vector<Point> sorted_points = merge_sort(points, 0);
    vector<Point> sorted_y = merge_sort(points, 1);
    auto result = closest_pair(sorted_points, sorted_y);
    cout << "Closest pair: ((" << result.first.first.first << ", " << result.first.first.second << "), ("
         << result.first.second.first << ", " << result.first.second.second << "))\n";
    cout << "Minimum distance: " << result.second << '\n';
    return 0;
}
